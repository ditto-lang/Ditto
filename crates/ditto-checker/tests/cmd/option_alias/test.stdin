module Test exports (..);

import (base) Maybe (Maybe);

type alias Option(a) = Maybe(a);

some = fn (a: a): Option(a) -> Maybe.Just(a);
none = fn (): Option(a) -> Maybe.Nothing;
