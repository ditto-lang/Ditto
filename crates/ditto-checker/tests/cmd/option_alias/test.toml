bin.name = "ditto-checker-testbin-check-module"
args = ["option_alias.ditto"]
fs.sandbox = true
