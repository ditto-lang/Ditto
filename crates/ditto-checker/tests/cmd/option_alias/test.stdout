{
  "module_name": [
    "Test"
  ],
  "exports": {
    "types": {
      "Option": {
        "Alias": {
          "doc_comments": [],
          "doc_position": 0,
          "kind": {
            "Function": {
              "parameters": [
                "Type"
              ]
            }
          },
          "aliased_type": {
            "type": "Call",
            "data": {
              "function": {
                "type": "Constructor",
                "data": {
                  "constructor_kind": {
                    "Function": {
                      "parameters": [
                        "Type"
                      ]
                    }
                  },
                  "canonical_value": {
                    "module_name": [
                      "base",
                      [
                        "Maybe"
                      ]
                    ],
                    "value": "Maybe"
                  },
                  "source_value": {
                    "module_name": null,
                    "value": "Maybe"
                  }
                }
              },
              "arguments": [
                {
                  "type": "Variable",
                  "data": {
                    "variable_kind": "Type",
                    "var": 0,
                    "source_name": "a"
                  }
                }
              ]
            }
          },
          "alias_variables": [
            0
          ]
        }
      }
    },
    "constructors": {},
    "values": {
      "none": {
        "doc_comments": [],
        "doc_position": 0,
        "value_type": {
          "type": "Function",
          "data": {
            "parameters": [],
            "return_type": {
              "type": "Call",
              "data": {
                "function": {
                  "type": "ConstructorAlias",
                  "data": {
                    "constructor_kind": {
                      "Function": {
                        "parameters": [
                          "Type"
                        ]
                      }
                    },
                    "canonical_value": {
                      "module_name": [
                        null,
                        [
                          "Test"
                        ]
                      ],
                      "value": "Option"
                    },
                    "source_value": {
                      "module_name": null,
                      "value": "Option"
                    },
                    "alias_variables": [
                      0
                    ],
                    "aliased_type": {
                      "type": "Call",
                      "data": {
                        "function": {
                          "type": "Constructor",
                          "data": {
                            "constructor_kind": {
                              "Function": {
                                "parameters": [
                                  "Type"
                                ]
                              }
                            },
                            "canonical_value": {
                              "module_name": [
                                "base",
                                [
                                  "Maybe"
                                ]
                              ],
                              "value": "Maybe"
                            },
                            "source_value": {
                              "module_name": null,
                              "value": "Maybe"
                            }
                          }
                        },
                        "arguments": [
                          {
                            "type": "Variable",
                            "data": {
                              "variable_kind": "Type",
                              "var": 0,
                              "source_name": "a"
                            }
                          }
                        ]
                      }
                    }
                  }
                },
                "arguments": [
                  {
                    "type": "Variable",
                    "data": {
                      "variable_kind": "Type",
                      "var": 0,
                      "source_name": "a"
                    }
                  }
                ]
              }
            }
          }
        }
      },
      "some": {
        "doc_comments": [],
        "doc_position": 1,
        "value_type": {
          "type": "Function",
          "data": {
            "parameters": [
              {
                "type": "Variable",
                "data": {
                  "variable_kind": "Type",
                  "var": 0,
                  "source_name": "a"
                }
              }
            ],
            "return_type": {
              "type": "Call",
              "data": {
                "function": {
                  "type": "Constructor",
                  "data": {
                    "constructor_kind": {
                      "Function": {
                        "parameters": [
                          "Type"
                        ]
                      }
                    },
                    "canonical_value": {
                      "module_name": [
                        "base",
                        [
                          "Maybe"
                        ]
                      ],
                      "value": "Maybe"
                    },
                    "source_value": null
                  }
                },
                "arguments": [
                  {
                    "type": "Variable",
                    "data": {
                      "variable_kind": "Type",
                      "var": 0,
                      "source_name": "a"
                    }
                  }
                ]
              }
            }
          }
        }
      }
    }
  },
  "types": {
    "Option": {
      "Alias": {
        "doc_comments": [],
        "type_name_span": {
          "start_offset": 66,
          "end_offset": 72
        },
        "kind": {
          "Function": {
            "parameters": [
              "Type"
            ]
          }
        },
        "aliased_type": {
          "type": "Call",
          "data": {
            "function": {
              "type": "Constructor",
              "data": {
                "constructor_kind": {
                  "Function": {
                    "parameters": [
                      "Type"
                    ]
                  }
                },
                "canonical_value": {
                  "module_name": [
                    "base",
                    [
                      "Maybe"
                    ]
                  ],
                  "value": "Maybe"
                },
                "source_value": {
                  "module_name": null,
                  "value": "Maybe"
                }
              }
            },
            "arguments": [
              {
                "type": "Variable",
                "data": {
                  "variable_kind": "Type",
                  "var": 0,
                  "source_name": "a"
                }
              }
            ]
          }
        },
        "alias_variables": [
          0
        ]
      }
    }
  },
  "constructors": {},
  "values": {
    "none": {
      "doc_comments": [],
      "name_span": {
        "start_offset": 133,
        "end_offset": 137
      },
      "expression": {
        "expression": "Function",
        "data": {
          "span": {
            "start_offset": 140,
            "end_offset": 173
          },
          "function_type": {
            "type": "Function",
            "data": {
              "parameters": [],
              "return_type": {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "ConstructorAlias",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            "Type"
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Test"
                          ]
                        ],
                        "value": "Option"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "Option"
                      },
                      "alias_variables": [
                        0
                      ],
                      "aliased_type": {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "Constructor",
                            "data": {
                              "constructor_kind": {
                                "Function": {
                                  "parameters": [
                                    "Type"
                                  ]
                                }
                              },
                              "canonical_value": {
                                "module_name": [
                                  "base",
                                  [
                                    "Maybe"
                                  ]
                                ],
                                "value": "Maybe"
                              },
                              "source_value": {
                                "module_name": null,
                                "value": "Maybe"
                              }
                            }
                          },
                          "arguments": [
                            {
                              "type": "Variable",
                              "data": {
                                "variable_kind": "Type",
                                "var": 0,
                                "source_name": "a"
                              }
                            }
                          ]
                        }
                      }
                    }
                  },
                  "arguments": [
                    {
                      "type": "Variable",
                      "data": {
                        "variable_kind": "Type",
                        "var": 0,
                        "source_name": "a"
                      }
                    }
                  ]
                }
              }
            }
          },
          "binders": [],
          "body": {
            "expression": "ImportedConstructor",
            "data": {
              "span": {
                "start_offset": 160,
                "end_offset": 173
              },
              "constructor_type": {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "ConstructorAlias",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            "Type"
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Test"
                          ]
                        ],
                        "value": "Option"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "Option"
                      },
                      "alias_variables": [
                        0
                      ],
                      "aliased_type": {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "Constructor",
                            "data": {
                              "constructor_kind": {
                                "Function": {
                                  "parameters": [
                                    "Type"
                                  ]
                                }
                              },
                              "canonical_value": {
                                "module_name": [
                                  "base",
                                  [
                                    "Maybe"
                                  ]
                                ],
                                "value": "Maybe"
                              },
                              "source_value": {
                                "module_name": null,
                                "value": "Maybe"
                              }
                            }
                          },
                          "arguments": [
                            {
                              "type": "Variable",
                              "data": {
                                "variable_kind": "Type",
                                "var": 0,
                                "source_name": "a"
                              }
                            }
                          ]
                        }
                      }
                    }
                  },
                  "arguments": [
                    {
                      "type": "Variable",
                      "data": {
                        "variable_kind": "Type",
                        "var": 0,
                        "source_name": "a"
                      }
                    }
                  ]
                }
              },
              "constructor": {
                "module_name": [
                  "base",
                  [
                    "Maybe"
                  ]
                ],
                "value": "Nothing"
              }
            }
          }
        }
      }
    },
    "some": {
      "doc_comments": [],
      "name_span": {
        "start_offset": 88,
        "end_offset": 92
      },
      "expression": {
        "expression": "Function",
        "data": {
          "span": {
            "start_offset": 95,
            "end_offset": 132
          },
          "function_type": {
            "type": "Function",
            "data": {
              "parameters": [
                {
                  "type": "Variable",
                  "data": {
                    "variable_kind": "Type",
                    "var": 0,
                    "source_name": "a"
                  }
                }
              ],
              "return_type": {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "Constructor",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            "Type"
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          "base",
                          [
                            "Maybe"
                          ]
                        ],
                        "value": "Maybe"
                      },
                      "source_value": null
                    }
                  },
                  "arguments": [
                    {
                      "type": "Variable",
                      "data": {
                        "variable_kind": "Type",
                        "var": 0,
                        "source_name": "a"
                      }
                    }
                  ]
                }
              }
            }
          },
          "binders": [
            [
              {
                "Variable": {
                  "span": {
                    "start_offset": 99,
                    "end_offset": 100
                  },
                  "name": "a"
                }
              },
              {
                "type": "Variable",
                "data": {
                  "variable_kind": "Type",
                  "var": 0,
                  "source_name": "a"
                }
              }
            ]
          ],
          "body": {
            "expression": "Call",
            "data": {
              "span": {
                "start_offset": 119,
                "end_offset": 132
              },
              "call_type": {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "Constructor",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            "Type"
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          "base",
                          [
                            "Maybe"
                          ]
                        ],
                        "value": "Maybe"
                      },
                      "source_value": null
                    }
                  },
                  "arguments": [
                    {
                      "type": "Variable",
                      "data": {
                        "variable_kind": "Type",
                        "var": 0,
                        "source_name": "a"
                      }
                    }
                  ]
                }
              },
              "function": {
                "expression": "ImportedConstructor",
                "data": {
                  "span": {
                    "start_offset": 119,
                    "end_offset": 129
                  },
                  "constructor_type": {
                    "type": "Function",
                    "data": {
                      "parameters": [
                        {
                          "type": "Variable",
                          "data": {
                            "variable_kind": "Type",
                            "var": 0,
                            "source_name": "a"
                          }
                        }
                      ],
                      "return_type": {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "Constructor",
                            "data": {
                              "constructor_kind": {
                                "Function": {
                                  "parameters": [
                                    "Type"
                                  ]
                                }
                              },
                              "canonical_value": {
                                "module_name": [
                                  "base",
                                  [
                                    "Maybe"
                                  ]
                                ],
                                "value": "Maybe"
                              },
                              "source_value": null
                            }
                          },
                          "arguments": [
                            {
                              "type": "Variable",
                              "data": {
                                "variable_kind": "Type",
                                "var": 0,
                                "source_name": "a"
                              }
                            }
                          ]
                        }
                      }
                    }
                  },
                  "constructor": {
                    "module_name": [
                      "base",
                      [
                        "Maybe"
                      ]
                    ],
                    "value": "Just"
                  }
                }
              },
              "arguments": [
                {
                  "Expression": {
                    "expression": "LocalVariable",
                    "data": {
                      "span": {
                        "start_offset": 130,
                        "end_offset": 131
                      },
                      "variable_type": {
                        "type": "Variable",
                        "data": {
                          "variable_kind": "Type",
                          "var": 0,
                          "source_name": "a"
                        }
                      },
                      "variable": "a"
                    }
                  }
                }
              ]
            }
          }
        }
      }
    }
  },
  "values_toposort": [
    "none",
    "some"
  ]
}